pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py

```diff
@@ -148,15 +151,36 @@ def _convert_one(self, option: dict[str, Any]) -> list[str]:
 
         return [ l for part in blocks for l in part ]
 
+    # this could return a TState parameter, but that does not allow dependent types and
+    # will cause headaches when using BaseConverter as a type bound anywhere. Any is the
+    # next best thing we can use, and since this is internal it will be mostly safe.
+    @abstractmethod
+    def _prepare_fork(self) -> Any: raise NotImplementedError()
+    # this should return python 3.11's Self instead to ensure that a prepare+finish
+    # round-trip ends up with an object of the same type. for now we'll use BaseConverter
+    # since it's good enough so far.
+    @classmethod
+    @abstractmethod
+    def _finish_fork(cls, a: Any) -> BaseConverter: raise NotImplementedError()
```

Comment:
Is there a reason we cannot just use a closure instead of passing the flags separately? Some kind of serialization issue? The following appears to work fine:

```patch
diff --git a/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py b/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py
index 087c98b3d3d..24dbccdc1d4 100644
--- a/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py
+++ b/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/options.py
@@ -155,13 +155,7 @@ class BaseConverter(Converter):
     # will cause headaches when using BaseConverter as a type bound anywhere. Any is the
     # next best thing we can use, and since this is internal it will be mostly safe.
     @abstractmethod
-    def _prepare_fork(self) -> Any: raise NotImplementedError()
-    # this should return python 3.11's Self instead to ensure that a prepare+finish
-    # round-trip ends up with an object of the same type. for now we'll use BaseConverter
-    # since it's good enough so far.
-    @classmethod
-    @abstractmethod
-    def _finish_fork(cls, a: Any) -> BaseConverter: raise NotImplementedError()
+    def _prepare_fork(self) -> Callable[[], BaseConverter]: raise NotImplementedError()
 
     def _render_option(self, name: str, option: dict[str, Any]) -> RenderedOption:
         try:
@@ -170,15 +164,11 @@ class BaseConverter(Converter):
             raise Exception(f"Failed to render option {name}") from e
 
     @classmethod
-    def _init_parallel_render(cls, arg: Any) -> Any:
-        return cls._finish_fork(arg)
-    @classmethod
     def _do_parallel_render(cls, s: Any, a: Any) -> Any:
         return s._render_option(*a)
 
     def add_options(self, options: dict[str, Any]) -> None:
-        mapped = parallel.map(self._do_parallel_render, options.items(), 100,
-                              self._init_parallel_render, self._prepare_fork())
+        mapped = parallel.map(self._do_parallel_render, options.items(), 100, self._prepare_fork())
         for (name, option) in zip(options.keys(), mapped):
             self._options[name] = option
 
@@ -218,12 +208,8 @@ class DocBookConverter(BaseConverter):
         self._varlist_id = varlist_id
         self._id_prefix = id_prefix
 
-    def _prepare_fork(self) -> Any:
-        return (self._manpage_urls, self._revision, self._markdown_by_default, self._document_type,
-                self._varlist_id, self._id_prefix)
-    @classmethod
-    def _finish_fork(cls, a: Any) -> DocBookConverter:
-        return cls(*a)
+    def _prepare_fork(self) -> Callable[[], BaseConverter]:
+        return lambda: self.__class__(self._manpage_urls, self._revision, self._markdown_by_default, self._document_type, self._varlist_id, self._id_prefix)
 
     def _render_code(self, option: dict[str, Any], key: str) -> list[str]:
         if lit := option_is(option, key, 'literalDocBook'):
@@ -321,11 +307,8 @@ class ManpageConverter(BaseConverter):
         self._options_by_id = _options_by_id or {}
         super().__init__({}, revision, markdown_by_default)
 
-    def _prepare_fork(self) -> Any:
-        return ((self._revision, self._markdown_by_default), { '_options_by_id': self._options_by_id })
-    @classmethod
-    def _finish_fork(cls, a: Any) -> ManpageConverter:
-        return cls(*a[0], **a[1])
+    def _prepare_fork(self) -> Callable[[], BaseConverter]:
+        return lambda: self.__class__((self._revision, self._markdown_by_default), { '_options_by_id': self._options_by_id })
 
     def _render_option(self, name: str, option: dict[str, Any]) -> RenderedOption:
         assert isinstance(self._md.renderer, OptionsManpageRenderer)
diff --git a/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/parallel.py b/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/parallel.py
index 1995b5306f8..18264608ecc 100644
--- a/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/parallel.py
+++ b/pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/parallel.py
@@ -15,20 +15,20 @@ pool_processes: Optional[int] = None
 class _Map:
     @classmethod
     def init(cls: Any, *args: Any) -> None:
-        (cls.fn, cls.state_fn, cls.state_arg) = args
+        (cls.fn, cls.state_fn) = args
 
     @classmethod
     def map(cls: Any, arg: Any, state: Any = []) -> Any:
         # if a Pool initializer throws it'll just be retried, leading to endless loops.
         # doing the proper initialization only on first use avoids this.
         if not state:
-            state.append(cls.state_fn(cls.state_arg))
+            state.append(cls.state_fn())
         return cls.fn(state[0], arg)
 
 def map(fn: Callable[[S, T], R], d: Iterable[T], chunk_size: int,
-        state_fn: Callable[[A], S], state_arg: A) -> list[R]:
+        state_fn: Callable[[A], S]) -> list[R]:
     """
-    `[ fn(state, i) for i in d ]`  where `state = state_fn(state_arg)`, but using multiprocessing
+    `[ fn(state, i) for i in d ]`  where `state = state_fn()`, but using multiprocessing
     if `pool_processes` is not `None`. when using multiprocessing is used the state function will
     be run once in ever worker process and `multiprocessing.Pool.imap` will be used.
 
@@ -36,7 +36,7 @@ def map(fn: Callable[[S, T], R], d: Iterable[T], chunk_size: int,
     discrepancies if `pool_processes` is not None, since each worker will have its own copy.
     """
     if pool_processes is None:
-        state = state_fn(state_arg)
+        state = state_fn()
         return [ fn(state, i) for i in d ]
-    with multiprocessing.Pool(pool_processes, _Map.init, (fn, state_fn, state_arg)) as p:
+    with multiprocessing.Pool(pool_processes, _Map.init, (fn, state_fn)) as p:
         return list(p.imap(_Map.map, d, chunk_size))
```

Reply:
Looks like “Explicitly pass resources to child processes” section of the [multiprocessing programming guidelines](https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming) applies.